### ISSUE TITLE
Couldn't reproduce iPQ quantization results -- Doesn't have Distillation Loss, huge performance drop

### ISSUE THREAD

**Issue by yilinyang7** [pytorch/fairseq#2340]
## 🐛 Bug

From your Quant-Noise paper, your methods would always require iPQ quantization at the last step, however, I couldn't reproduce your reported performance on iPQ quantization.
For example, you report iPQ quantization on finetuned RoBERTa could reach 82.5 on MNLI, but I'm not able to reproduce this.
I looked into the code, and it seems the biggest difference is that iPQ would fine-tune the centroids on distillation loss against the uncompressed model, but your code doesn't implement this, and finetuned only on golden targets.

Is it the case that you have it implemented in your internal branch, and is it possible to publish that essential part of the code to reproduce your reported performance?

### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. Fine-tuning RoBERTa model on MNLI:
`fairseq-train MNLI-bin --restore-file pretrained_model/roberta.base/model.pt --max-positions 512 --max-sentences 32 --max-tokens 4400 --task sentence_prediction --reset-optimizer --reset-dataloader --reset-meters --required-batch-size-multiple 1 --init-token 0 --separator-token 2 --arch roberta --criterion sentence_prediction --num-classes 3 --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.1 --optimizer adam --adam-betas "(0.9, 0.98)" --adam-eps 1e-06 --clip-norm 0.0 --lr-scheduler polynomial_decay --lr 1e-5 --total-num-update 123873 --warmup-updates 7432 --fp16 --fp16-init-scale 4 --threshold-loss-scale 1 --fp16-scale-window 128 --max-epoch 10 --find-unused-parameters --best-checkpoint-metric accuracy —maximize-best-checkpoint-metric —save-dir checkpoints/roberta-base-mnli`
2. run iPQ quantization on finetuned RoBERTa model
`fairseq-train --task sentence_prediction MNLI-bin --restore-file checkpoints/roberta-base-mnli/checkpoint_best.pt --save-dir checkpoints/roberta-base-mnli-ipq1 --reset-optimizer --reset-dataloader --reset-meters --max-positions 512 --max-sentences 32 --max-tokens 4400 --max-update 75714 --warmup-updates 4536 --lr 1e-5 --init-token 0 --separator-token 2 --arch roberta --criterion sentence_prediction --num-classes 3 --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.1 --optimizer adam --adam-betas "(0.9, 0.98)" --adam-eps 1e-06 --clip-norm 0.0 --lr-scheduler polynomial_decay --fp16 --fp16-init-scale 4 --threshold-loss-scale 1 --fp16-scale-window 128 --no-progress-bar --skip-invalid-size-inputs-valid-test —quantization-config-path examples/quant_noise/transformer_quantization_config.yaml`

Best wishes.

**Comment by yilinyang7** [comment]
sorry for directly @-ing, i think it's most related to @huihuifan

**Comment by pierrestock** [comment]
Hi yilinyang7, 

Thanks a lot for reaching out! Our results in the paper are implemented using "golden targets" (that is, 1-hot encodings). We did not use, for RoBERTa, any kind of distillation loss. 

Therefore, the command should output the good accuracy. Could you give us more precisions on what the scripts outputs, both for steps 1 and 2? 

Thanks,

Pierre

**Comment by yilinyang7** [comment]
Hello @pierrestock,

Thank you for your reply!
Yes indeed. The first step of RoBERTa finetuning on MNLI does give reported performance, around 87 F1 score. However, the second step of iPQ quantization & centroids finetuning could only reach around 75 F1 score on MNLI, which is significantly lower than reported. Besides accuracy, I'm able to get reported compression ratio: around 20x smaller (447.9M -> 25M).

More details on the replication: I loosely follows the command provided [here](https://github.com/pytorch/fairseq/blob/master/examples/quant_noise/README.md#iterative-product-quantization-1), which actually failed to provide lr or max-steps.
Since the current quantization_config file isn't runnable, I modify it as follows (I basically revise "layers_to_quantize" to their correct terms):
```
# Number of Centroids for Product Quantization, by default 256 (byte-aligned)
n_centroids:
    Linear:
        key: in_features
        value: {"*": 256}
    Embedding:
        key: embedding_dim
        value: {"*": 256}

# Block Sizes for Product Quantization
# We suggest: 8 for FFN, 4 for ATTN, 4 for embedding projections, 8 for embeddings
block_sizes:
  Linear:
      key: fuzzy_name
      value: {fc: 8, attn: 4, emb: 4}
  Embedding:
      key: fuzzy_name
      value: {emb: 8}

# Layers to Quantize Sequentially
# We suggest: first FFN, then EMB, then ATTN
layers_to_quantize:
    - encoder\.sentence_encoder\.layers\.\d+\.fc[12]
    - encoder\.sentence_encoder\.embed_tokens
    - encoder\.sentence_encoder\.layers\.\d+\.self_attn\.(k_proj|v_proj|q_proj|out_proj)
```
For iPQ quantization, I run it only on one gpu, with few hyper-parameters I tried to tune:
--max-update 75714 --warmup-updates 4536

I also try adapting Quant-Noise in the first step of RoBERTa finetuning, but its iPQ-ed model would still reach similar performance of around 75 F1 score.

Let me know your thoughts, I'm pretty thrilled to see such compression ratio of sota RoBERTa models.

Best,
Yilin

**Comment by yilinyang7** [comment]
Hi @pierrestock,

Do you mind sharing your commands to replicate iPQ quantization and centroids finetuning, so that I could evaluate the difference?

Best.

**Comment by yilinyang7** [comment]
Hello authors,

I tried finetuning centroids for more epochs (12 in total), and I'm able to reach ~80.5 F1 score on MNLI, which is still lower than reported (82.5 F1). I have to say I couldn't replicate your result from your codes. If you find anything, please let me know.

Thanks.

**Comment by wangyihi** [comment]
HI @pierrestock,
Do you mind sharing your model with trained on RoBERTa+QuantNoise? 
 Thanks

**Comment by pierrestock** [comment]
Hi yilinyang and wangyihi,

First of all, sorry for the answer lag, I was switching jobs. On the top of my mind (from what I recall), a few things that could help here:
- Finetuning the centroids longer with a decaying learning rate
- Finetuing order of the types of layers (first FFNs, then attentions)

cc @huihuifan for more input
